protocol/rpcconsumer/consumer_relay_state_machine.go

```diff
@@ -83,24 +111,72 @@ func (crsm *ConsumerRelayStateMachine) GetSelection() Selection {
 	return crsm.selection
 }
 
-func (crsm *ConsumerRelayStateMachine) shouldRetryOnResult(numberOfRetriesLaunched int, numberOfNodeErrors uint64) bool {
-	shouldRetry := crsm.shouldRetryInner(numberOfRetriesLaunched)
-	// archive functionality will be added here.
+// Should retry implements the logic for when to send another relay.
+// As well as the decision of changing the protocol message,
+// into different extensions or addons based on certain conditions
+func (crsm *ConsumerRelayStateMachine) shouldRetry(numberOfRetriesLaunched int, numberOfNodeErrors uint64) bool {
+	shouldRetry := crsm.retryCondition(numberOfRetriesLaunched)
+	if shouldRetry {
+		// Retry archive logic
+		hashes := crsm.GetProtocolMessage().GetRequestedBlocksHashes()
+		if len(hashes) > 0 && numberOfNodeErrors > 0 {
+			// Launch archive only on the second retry attempt.
+			if numberOfRetriesLaunched == 1 {
+				// Iterate over all hashes found in relay, if we don't have them in the cache we can try retry on archive.
+				// If we are familiar with all, we don't want to allow archive.
+				for _, hash := range hashes {
+					if !crsm.relayRetriesManager.CheckHashInCache(hash) {
+						// If we didn't find the hash in the cache we can try archive relay.
+						relayRequestData := crsm.protocolMessage.RelayPrivateData()
```

Comment:
in the StateMachine you should manage state per batch instead of checking it all the time via arguments, this will simplify the code and make debugging easier

the code should look like:
type RelayState struct {
    isArchive
    isUpgraded
    batch
    providersInBatch // maybe can help?
    routerKey // maybe a better place for it than single consumer session, and it holds the RelayState object
    protocolMessage // might be a nice container for protocolMessage
}
statemachine[batchNumber].IsArchive()
statemachine[batchNumber].IsUpgraded()

before you advance a batch, you set the new state, then when the state transition happens you do the things, not in the should retry function
then all the stuff happen on state transition, you can hook specific things like transition form regular relay to archive one
you set the state transition function to receive the two states and handle them, you can set good unitests